synapse/storage/databases/main/relations.py

```diff
@@ -765,6 +765,41 @@ def _get_if_user_has_annotated_event(txn: LoggingTransaction) -> bool:
             "get_if_user_has_annotated_event", _get_if_user_has_annotated_event
         )
 
+    async def get_mutual_event_relations(
```

Comment:
This could potentially pull a lot of data for a single event, but the alternatives I can think of are:

* Pre-process the events list to figure out what combinations of relation type / sender / event types are even of interest (but this might just cause a more complicated query?)
* Query (with a cache) when we reach a specific  `relation_match` rule, which would be many small queries.

Any thoughts on if these are better?

Reply:
Yeah, I don't like the fact that could take out an unbounded number of rows, especially for long running threads.

One alternative is to have a table that tracks this information, i.e. a table from `parent_id -> (rel_type, event_type, sender)`. I think we'd also want to pre-process the push rules to get all sender/rel types we care about and fetch them in one go and cache that maybe? Or something? 

Broadly I don't want to have to have the DB potentially walk an unbounded number of rows in `event_relations` table.

Reply:
> Isn't this pretty much the `event_relations` table + the `JOIN` I did above? I'm not sure what having a separate table would buy us.

The difference is that the new table would be bounded by the number of users participating in the thread, rather than the number of messages in the thread.

Reply:
> > Isn't this pretty much the `event_relations` table + the `JOIN` I did above? I'm not sure what having a separate table would buy us.
> 
> The difference is that the new table would be bounded by the number of users participating in the thread, rather than the number of messages in the thread.

Right, because it is only the unique rows. That seems quite heavy for this feature. 😢 

Adding the `DISTINCT` to the query does bound the number of *returned* rows, but not the number of walked rows. Will think a bit more about whether searching for specific tuples of information seems more efficient.

Reply:
Arg, the issue with this is that push rules can take glob-patterns. I guess we could make it so that the relation type can't do that, but I think that would be unique to this single rule kind then?

What do you think @erikjohnston?